### PR TITLE
Simplified Form

### PR DESCRIPTION
**Pull request by jaredreich**
General:
- [x] Handle prop changes (used components individually outside of Form)
- [x] Parse files from form (FormData)
- [ ] Optimize re-renders
- [x] Add demo for custom form field (eg. date)
- [ ] Update docs
- [x] Cleanup code
- [x] Ensure defaultValue and validation is working for all the components
- [x] Ensure updating is working for all components

Components:
- [x] Checkbox
- [x] CheckboxGroup
- [x] FileInput
- [x] ImageInput
- [x] DateInput
- [x] Input
- [x] Radio
- [x] RadioGroup
- [x] Select
- [x] Switch
- [x] Textarea

Components working individually:
- [x] Checkbox
- [x] CheckboxGroup
- [x] FileInput
- [x] ImageInput
- [x] DateInput
- [x] Input
- [x] Radio
- [x] RadioGroup
- [x] Select
- [x] Switch
- [x] Textarea

### CODE REVIEW

**Comment by jaredreich** [comment]
@stevenyuen can you take a peek at this, specifically the the new DateInput component. Let me know what you think and what kind of style improvements you'd make.

File: demo/content/Form.js [hunk 1]
```diff
@@ -38,8 +39,18 @@ export default class DemoContentForm extends Component {
...
    }
 
-   handleSubmit(data, files, formData) {
-      console.log(data, files) // eslint-disable-line no-console
+   handleSubmit(data, files, formData, constructFormData) {
+      /* eslint-disable */
+      console.log(data)
+      console.log(files)
+      const formdata = {}
+      for (const pair of formData.entries()) formdata[pair[0]] = pair[1]
```
Comment (asabhaney) [inline demo/content/Form.js]: `FormData#entries` (and most other `FormData` methods) are not supported on many browsers, including IE Edge & Safari. See https://developer.mozilla.org/en/docs/Web/API/FormData#Browser_compatibility

File: demo/content/Form.js [hunk 2]
```diff
@@ -84,7 +95,6 @@ export default class DemoContentForm extends Component {
                                  name="name.first"
                                  label="First Name"
                                  placeholder="Please enter your first name"
-                                 value="Jared"
                                  rules={['required']}
```
Comment (asabhaney) [inline demo/content/Form.js]: Any reason you removed all examples of `defaultValue` and `value`? Could we add a couple examples using this?

File: src/components/Form/CheckboxGroup/index.js [hunk 3]
```diff
@@ -1,73 +1,121 @@
...
-         this.setState({ value: newValue })
+   componentDidMount() {
+      if (this.context.OIOForm && this.props.name) {
+         this.context.OIOForm.setDefaultValue(this.props.name, this.state.value)
+         this.context.OIOForm.setRules(this.props.name, this.props.rules)
       }
    }
 
+   componentWillReceiveProps(nextProps) {
+      // TODO: If name changes, need to remove form value corresponding to old name
```
Comment (asabhaney) [inline src/components/Form/CheckboxGroup/index.js]: Has this been handled somewhere? (and in every other component?)

File: demo/content/Form.js (hunk 1 above)
Comment (jaredreich) [inline demo/content/Form.js]: This is fixed now

File: demo/content/Form.js (hunk 2 above)
Comment (jaredreich) [inline demo/content/Form.js]: Yay! Re-added 8a6f4fc

File: src/components/Form/CheckboxGroup/index.js (hunk 3 above)
Comment (jaredreich) [inline src/components/Form/CheckboxGroup/index.js]: Leaving as TODO for now as per discussion

**Comment by jaredreich** [comment]
@asabhaney back at you for review

**Comment by asabhaney** [comment]
@jaredreich Seems that sometimes the term `defaultValue` is used and other times `initialValue` is used. Would it be a good idea to choose one term and be consistent? I'd maybe lean towards `initialValue` since `defaultValue` already kind of means something in React form elements, so that could be a bit confusing?

**Comment by jaredreich** [comment]
@asabhaney shall we totally eliminate the usage of defaultValue? Currently, it is used to set a default value that is never async... but you’re absolutely right, perhaps it can all be done with initialValue

**Comment by jaredreich** [comment]
@asabhaney in this commit 7567d00, all usage of `defaultValue` was removed. Everything works with only `initialValue` now.

File: demo/content/Form.js [hunk 4]
```diff
@@ -27,19 +28,37 @@ export default class DemoContentForm extends Component {
       contents: React.PropTypes.array
    }
 
-   constructor(props) {
-      super(props)
+   constructor(props, context) {
+      super(props, context)
```
Comment (stevenyuen) [inline demo/content/Form.js]: context not required

Comment (asabhaney) [inline demo/content/Form.js]: Good catch. Addressed in https://github.com/mother/oio/pull/83.

**Comment by asabhaney** [comment]
@stevenyuen Back to you

**Comment by asabhaney** [comment]
One outstanding issue for `pristine` calculation, will address later. See #89.

**Pull request merged**